README.md
# tr-55

A Python implementation of a modified TR-55 stormwater runoff model with STEP-L like water quality routines.


## Installation

You can install the latest version from PyPI
```bash
pip install tr55
```

The `simulate_day` is the function most likely to be of direct interest for users of this module.

The `simulate_water_quality` and `simulate_modifications` functions are two other functions which can be used to create simulations with additional behaviors beyond those supplied by `simulate_day`.


## `simulate_day`

This function takes four arguments: a census of the area of interest (see the description given below in the discussion of `simulate_modifications`), an amount of precipitation in inches, an optional cell resolution (the size of a cell in square meters), and an optional Boolean  to control whether or not a Pre-Columbian simulation is done.

For each cell type present in the area of interest, it calculates runoff, infiltration, evapotranspiration, and pollutant loads caused by that cell type.  The algorithm used to calculate the water volumes is close to TR-55, the algorithm found in [the USDA's Technical Release 55, revised 1986](http://www.cpesc.org/reference/tr55.pdf), but with a few differences.  The main difference is the use of *Pitt Small Storm Hydrology Model* for low levels of precipitation when the land use is a built-type.  STEP-L like routines are used for the water quality calculations.

## Functions for Custom Scenarios

### `simulate_water_quality`

The `simulate_water_quality` function does a water quality calculation over an entire area of interest.  The arguments are:

   1. `tree`, the tree-like dictionary which contains the distribution of cell types in the area of interest.  For example:

   ```Python
   {
       "cell_count": 8,
       "distribution": {
           "c:developed_high": {"cell_count": 5},
           "a:deciduous_forest": {
               "cell_count": 3
               "distribution": {
                   "a:deciduous_forest": {"cell_count": 1},
                   "a:no_till": {"cell_count": 1},
                   "d:barren_land": {"cell_count", 1}
               }
           }
       }
   }
   ```

   represents an area of interest that is eight cells in size, with five of those cells of type `"c:developed_high"` (*Developed High Intensity* land use on top of type *C* soil), and one cell each of *deciduous forest*, *no-till farming*, and *barren land*.

   The single cells of *deciduous forest*, *no-till*, and the *barren land* are all underneath a node of three cells of type *deciduous forest*.  That indicates a land use modification has taken place: in this case, two of three original cells of *deciduous forest* have undergone modifications.

   2. The `cell_res` parameter gives the resolution (size) of each cell.  It is used for converting runoff, evapotranspiration, and infiltration amounts from inches to volumes.

   3. `fn` is the function that is used to perform the runoff, evapotranspiration, and infiltration calculation.  It takes `cell` and `cell_count` as arguments.

   4. The optional parameter `current_cell` contains the string description of the cell currently being worked on (e.g. "a:deciduous_forest").

   5. The optional Boolean `precolumbian` determines whether to simulate the cell type as-shown or under Pre-Columbian circumstances.  When a Pre-Columbian simulation is done, all developed land (NLCD 21-24) uses are treated as *mixed forest*.

In all probability, the fourth parameter will not need to be supplied if you are calling this function from external code.

### `simulate_modifications`

This function is used to simulate the effects of land use modifications.  The arguments are:

   1. `census` contains the distribution of cell types in the area of interest, along with an array of modifications.  For example, the following:

   ```Python
   {
       "cell_count": 8,
       "distribution": {
           "c:developed_high": {"cell_count": 5},
           "a:deciduous_forest": {"cell_count": 3}
       },
       "modifications": [
           {
               "change": "::no_till",
               "cell_count": 1,
               "distribution": {
                   "a:deciduous_forest": {"cell_count": 1},
               }
           },
           {
               "change": "d:barren_land:",
               "cell_count": 1,
               "distribution": {
                   "a:deciduous_forest": {"cell_count": 1}
               }
           }
       ]
   }
   ```

   is the census that corresponds to the `tree` given in the discussion of `simulate_water_quality` above.  There is an area of interest eight cells in size, with five of type `"c:developed_high"` and three of type `"a:deciduous_forest"`.

   Modifications are given as an array of dictionaries.  Each dictionary contains a `change` key whose value encodes the modification that has taken place.  In the example above, `"::no_till"` indicates that the no-till farming BMP has been applied, while `"a:barren_land:"` means that that particular area has been reclassified as being most barren_land sitting on top of A-type soil.

   2. The `fn` argument is as described previously in the discussion of `simulate_water_quality`.  It is responsible for performing the simulation at each cell.

   3. The `cell_res` argument is as described previously.

   4. The optional Boolean parameter `precolumbian` determines whether to simulate the cell type as-shown or under Pre-Columbian circumstances.  When a Pre-Columbian simulation is done, all land uses other than *water* and *wetland* are treated as *mixed forest*.

The output is dictionary with two keys, `modified` and `unmodified`.  These respectively contain modified and unmodified trees (the trees are as described in the discussion of `simulate_water_quality`) with runoff, evapotranspiration, infiltration, and pollutant loads included.


## Allowed Types

The following land use values are implemented and correspond to the keys listed:

 - 'open_water' - NLCD Type 11, Open Water
 - 'developed_open' - NLCD Type 21, Developed, Open Space
 - 'developed_low'- NLCD Type 22, Developed, Low Intensity
 - 'developed_med' - NLCD Type 23, Developed, Medium Intensity
 - 'developed_high' - NLCD Type 24, Developed, High Intensity
 - 'barren_land' - NLCD Type 31, Barren Land (Rock/Sand/Clay)
 - 'deciduous_forest' - NLCD Type 41, Deciduous Forest
 - 'evergreen_forest' - NLCD Type 42, Evergreen Forest
 - 'mixed_forest' - NLCD Type 43, Mixed Forest
 - 'shrub' - NLCD Type 52, Shrub/Scrub
 - 'grassland' - NLCD Type 71, Grassland/Herbaceous
 - 'pasture' - NLCD Type 81, Pasture/Hay
 - 'cultivated_crops' - NLCD Type 82, Cultivated Crops
 - 'woody_wetlands' - NLCD Type 90, Woody Wetlands
 - 'herbaceous_wetlands' - NLCD Type 95, Emergent Herbaceous Wetlands


## Usage Example

The output of the following program:
```Python
# -*- coding: utf-8 -*-
from __future__ import print_function
from __future__ import unicode_literals
from __future__ import division

import pprint

from tr55.model import simulate_day

census = {
    "cell_count": 147,
    "distribution": {
        "d:developed_med": {"cell_count": 33},
        "c:developed_high": {"cell_count": 42},
        "a:deciduous_forest": {"cell_count": 72}
    },
    "modifications": [
        {
            "change": "::no_till",
            "cell_count": 30,
            "distribution": {
                "d:developed_med": {"cell_count": 10},
                "c:developed_high": {"cell_count": 20}
            }
        },
        {
            "change": "d:barren_land:",
            "cell_count": 5,
            "distribution": {
                "a:deciduous_forest": {"cell_count": 5}
            }
        }
    ]
}

pprint.pprint(simulate_day(census, 0.984))
```
is partially reproduced here:
```Python
u'unmodified': {u'bod': 43.11309178874012,
                 u'cell_count': 147,
                 u'distribution': {u'a:deciduous_forest': {u'bod': 0.0,
                                                           u'cell_count': 72,
                                                           u'et': 0.14489999999999997,
                                                           u'inf': 0.8391,
                                                           u'runoff': 0.0,
                                                           u'tn': 0.0,
                                                           u'tp': 0.0,
                                                           u'tss': 0.0},
                                   u'c:developed_high': {u'bod': 28.342317499361275,
                                                         u'cell_count': 42,
                                                         u'et': 0.012322664565942195,
                                                         u'inf': 0.0,
                                                         u'runoff': 0.9716773354340579,
                                                         u'tn': 0.2079960397130545,
                                                         u'tp': 0.03291365903151632,
                                                         u'tss': 5.780461367410053},
                                   u'd:developed_med': {u'bod': 14.770774289378844,
                                                        u'cell_count': 33,
                                                        u'et': 0.037259999999999995,
                                                        u'inf': 0.26946506358880085,
                                                        u'runoff': 0.6772749364111992,
                                                        u'tn': 0.08574559651037718,
                                                        u'tp': 0.014395246129479379,
                                                        u'tss': 1.764982351527472}},
                 u'et': 0.08285667967190184,
                 u'inf': 0.47147991223422064,
                 u'runoff': 0.4296634080938776,
                 u'tn': 0.2937416362234317,
                 u'tp': 0.0473089051609957,

CHANGELOG.md
## 1.3.0

- Improved Water Quality event mean concentrations per updated sources from Barry Evans


## 1.2.1

- Restore perennial ice land use option

## 1.2.0

- Added lots of information into tables.py about all of the data sources for the values
- Improved curve numbers for TR55 (NRCS) for developed lands
- Rewrote Pitt's Small Storm Hydrology model to correctly be soil-type dependent and to use a look-up table for values instead of a polynomial derived from that look-up table
- For the new implementation of the Pitt/SSH model, numpy is now a requirement
- Improved evapotranspiration crop coefficients for all except developed lands. Developed Ki's still need to be sourced and fixed.
- Added a limit to the amount of runoff that can be converted to infiltration for infiltration-type BMP's. This is based on the amount of rainfall that lands on the BMP and an artificial ratio of contributing area to BMP area. Note that this ratio is based only on recommendations by the PA DEP and has nothing to do with the actual contributing areas for the BMPs as drawn.
- Recreated cluster housing as an independent land use rather than using numbers for developed_low.
- Rewrote all test modules to reflect all of the above changes.

## 1.1.3

- Fix bug which elevated the Green Roof BMP type storage capacity.  Aligned with provided documentation.

## 1.1.2

- Fix bug which threw an exception when runoff volume was 0 and BMPs were applied (essentially a precipitation value of 0).

## 1.1.1

- Allow infiltration to be increased during BMP modifications

## 1.1.0

- Allow rain garden, veg basin, porous pavement and green roof BMP types to reduce runoff
  over entire AoI according to an available volume for water storage per unit BMP Area

## 1.0.6

- Updated Curve Numbers for NLCD classes 21, 90 & 95 from input from Stroud Water Research
- Fixed Pitts impervious and urban grass calculations to align with published amounts
- Reduced imperviousness of `developed_high` NLCD type

## 1.0.4-1.0.5

- Update Exception types

## 1.0.3

- Removed unused code paths (`simulate_year`)
- Updated documentation
- Update correct D Soil type values for BMPs
- Update `barren_land` curve numbers to align with published model description

## 1.0.1-1.0.2

- Same as `1.0.0`, but with a valid `setup.py` version.

## 1.0.0

- Input format was changed. Strings used to run a model now align to NLCD types. A full list is provided in the readme. Other land types that were similar to NLCD types (and used these types to do calculations) were removed.
- Output strings were changed to match NLCD types.

## 0.2.2

- Same as `0.2.1`, but with a valid `setup.py` version.

## 0.2.1

- Fixes a bug in the computation of modified censuses. (https://github.com/WikiWatershed/tr-55/pull/40)

## 0.2.0

- The output format was changed. Previously, only the overall result of the calculation was given, after this change the overall result plus the sub-results for each (nlcd,soil) pair are given. (https://github.com/WikiWatershed/tr-55/pull/30)
- Some code was added to generate and compare mini-app ables. (https://github.com/WikiWatershed/tr-55/pull/31, https://github.com/WikiWatershed/tr-55/pull/38)
- Attempt to plug in the correct NLCD codes. Previously there were some "names" in the mini-app that did not have an associated NLCD code in the TR-55. (https://github.com/WikiWatershed/tr-55/pull/32)
- Changed the the format of the input expected by the code. It allows for areas that are covered by both reclassifications and BMPs. (https://github.com/WikiWatershed/tr-55/pull/35)
- Made it easier to do Pre-Columbian calculations. (https://github.com/WikiWatershed/tr-55/pull/37)

## 0.1.0

- Added simple STEP-L based water quality routines.

## 0.1.0.dev2

- Initial development release.


tr55/tables.py
# -*- coding: utf-8 -*-
from __future__ import print_function
from __future__ import unicode_literals
from __future__ import division

"""
TR-55 tables
"""


# For the different land uses, this describes the NLCD class value, the landscape factor (ki) and the Curve
# Numbers for each hydrologic soil group for that use type.
# NOTE: Missing NLCD type 12 (plus all Alaska only types (51, 72-74))
# For the BMP's the numbers are not Curve Numbers, they are quantities of rainfall (in inches)
#  that will be converted to infiltration by that BMP for that soil type.

# The Food and Agriculture Organization of the United Nations (FAO) document on evapotranspiration is:
# Allen, R.G.; Pereira, L.S.; Raes, D.; Smith, M. Evapotranspiration and Crop Water Requirements;
# Irrigation and Drainage Paper No. 56; FAO: Rome, 1998.
# Available:  http://www.fao.org/docrep/x0490e/x0490e00.htm#Contents

LAND_USE_VALUES = {

        # NRCS Curve Numbers for NLCD land classes
    'open_water':           {'nlcd': 11, 'ki': 0.6525, 'cn': {'a': 100, 'b': 100, 'c': 100, 'd': 100}},
        # Curve Number Source:  Assumes 100% runoff
        # Ki Source: FAO for Open Water, > 5 m depth, clear of turbidity, temperate climate.
    'perennial_ice':        {'nlcd': 12, 'ki': 0.0, 'cn': {'a': 100, 'b': 100, 'c': 100, 'd': 100}},
        # Curve Number Source:  Assumes 100% runoff
        # Ki Source: Assumes no ET.
    'developed_open':       {'nlcd': 21, 'ki': 0.95, 'cn': {'a': 59, 'b': 75, 'c': 83, 'd': 87}},
        # Curve Number Source:  Blend of Pasture - medium and paved parking assuming 20% impervious.
            # (TR-55, 1986, Table 2-2a)
        # Ki Source: FAO for growing season for cool season turfgrass (dense stands of bluegrass, ryegrass, and fescue).
    'developed_low':        {'nlcd': 22, 'ki': 0.42, 'cn': {'a': 68, 'b': 80, 'c': 86, 'd': 89}},
        # Curve Number Source:  Blend of Pasture - medium and paved parking assuming 38% impervious.
            # (TR-55, 1986, Table 2-2a)
        # Ki Source: UNKNOWN
    'developed_med':        {'nlcd': 23, 'ki': 0.18, 'cn': {'a': 81, 'b': 88, 'c': 91, 'd': 93}},
        # Curve Number Source:  Blend of Pasture - medium and paved parking assuming 65% impervious.
            # (TR-55, 1986, Table 2-2a)
        # Ki Source: UNKNOWN
    'developed_high':       {'nlcd': 24, 'ki': 0.06, 'cn': {'a': 91, 'b': 94, 'c': 95, 'd': 96}},
        # Curve Number Source:  Blend of Pasture - medium and paved parking assuming 85% impervious.
        # Ki Source: UNKNOWN
    'barren_land':          {'nlcd': 31, 'ki': 0.30, 'cn': {'a': 77, 'b': 86, 'c': 91, 'd': 94}},
        # Curve Number Source:  Fallow, Bare soil; Newly graded areas (TR-55, 1986, Table 2-2a and 2-2b)
        # Ki Source: Sridhar, Venkataramana, "Evapotranspiration Estimation and Scaling Effects Over The Nebraska Sandhills"
        # (2007). Great Plains Research: A Journal of Natural and Social Sciences. Paper 870.
        # http://digitalcommons.unl.edu/greatplainsresearch/870
    'deciduous_forest':     {'nlcd': 41, 'ki': 1.0, 'cn': {'a': 30, 'b': 55, 'c': 70, 'd': 77}},
        # Curve Number Source:  Woods, Good condition;
            # Woods are protected from grazing and litter and brush adequately cover the soil.
            # (TR-55, 1986, Table 2-2c)
        # Ki Source: Sridhar, Venkataramana, "Evapotranspiration Estimation and Scaling Effects Over The Nebraska Sandhills"
        # (2007). Great Plains Research: A Journal of Natural and Social Sciences. Paper 870.
        # http://digitalcommons.unl.edu/greatplainsresearch/870
    'evergreen_forest':     {'nlcd': 42, 'ki': 1.00, 'cn': {'a': 30, 'b': 55, 'c': 70, 'd': 77}},
        # Curve Number Source:  Woods, Good condition;
            # Woods are protected from grazing and litter and brush adequately cover the soil.
            # (TR-55, 1986, Table 2-2c)
        # Ki Source: FAO for conifer trees during growing season in well-watered conditions for large forests.
    'mixed_forest':         {'nlcd': 43, 'ki': 1.0, 'cn': {'a': 30, 'b': 55, 'c': 70, 'd': 77}},
        # Curve Number Source:  Woods, Good condition;
            # Woods are protected from grazing and litter and brush adequately cover the soil.
            # (TR-55, 1986, Table 2-2c)
        # Ki Source: Sridhar, Venkataramana, "Evapotranspiration Estimation and Scaling Effects Over The Nebraska Sandhills"
        # (2007). Great Plains Research: A Journal of Natural and Social Sciences. Paper 870.
        # http://digitalcommons.unl.edu/greatplainsresearch/870
    'shrub':                {'nlcd': 52, 'ki': 0.90, 'cn': {'a': 35, 'b': 56, 'c': 70, 'd': 77}},
        # Curve Number Source:  Brush, fair; 50-75% ground cover (TR-55, 1986, Table 2-2c)
        # Ki Source: Descheemaeker, K., Raes, D., Allen, R., Nyssen, J., Poesen, J., Muys, B., Haile, M. and Deckers, J.
        # 2011. Two rapid appraisals of FAO-56 crop coefficients for semiarid natural vegetation of the
        # northern Ethiopian highlands. Journal of Arid Environments 75(4):353-359.
    'grassland':            {'nlcd': 71, 'ki': 1.08, 'cn': {'a': 30, 'b': 58, 'c': 71, 'd': 78}},
        # Curve Number Source:  Meadow - continuous grass, protected from grazing and generally mowed for hay.
            # (TR-55, 1986, Table 2-2c)
        # Ki Source: Average of all values in FAO document for Forages/Hay.
    'pasture':              {'nlcd': 81, 'ki': 0.95, 'cn': {'a': 39, 'b': 61, 'c': 74, 'd': 80}},
        # Curve Number Source:  Pasture, good; >75% ground cover and not heavily grazed. (TR-55, 1986, Table 2-2c)
        # Ki Source: FAO for Grazing pasture with rotated grazing.
    'cultivated_crops':     {'nlcd': 82, 'ki': 1.15, 'cn': {'a': 67, 'b': 78, 'c': 85, 'd': 89}},
        # Curve Number Source:  Row crops, straight rows, good condition (TR-55, 1986, Table 2-2b)
        # Ki Source: FAO average for all cereal crows during the growing season.
    'woody_wetlands':       {'nlcd': 90, 'ki': 1.20, 'cn': {'a': 30, 'b': 30, 'c': 30, 'd': 30}},
        # Curve Number Source: Uses lowest curve numbers possible to maximize infiltration
        # Ki Source: FAO for either Cattail/Bulrush wetland or Reed Swamp wetland during growing season.
    'herbaceous_wetlands':  {'nlcd': 95, 'ki': 1.20, 'cn': {'a': 30, 'b': 30, 'c': 30, 'd': 30}},
        # Curve Number Source: Uses lowest curve numbers possible to maximize infiltration
        # Ki Source: FAO for either Cattail/Bulrush wetland or Reed Swamp wetland during growing season.

    # NRCS Curve Numbers for BMP's acting as land cover changes
    'cluster_housing':      {'ki': 0.42, 'cn': {'a': 62, 'b': 77, 'c': 84, 'd': 88}},
        # Curve Number Source:  Blend of Pasture - medium and paved parking assuming 26.8% impervious.
        # Ki Source: UNKNOWN
    'no_till':              {'ki': 0.9, 'cn': {'a': 57, 'b': 73, 'c': 82, 'd': 86}},
        # Curve Number Source:  UNKNOWN
        # Ki Source:  UNKNOWN

    # Storage Capacities and Maximum Loading Ratios for Infiltration BMP's
    # storage is in m3/m2, max_drainage_ratio is the ratio of drawn BMP area to
    # the maximum possible area that should contribute to it.
    # NOTE that these contributing areas ratios are based only on the suggested
    # drainage areas for a well-designed BMP's and have nothing to do with the
    # user's actual placement of the BMP on the UI map.
    'green_roof':           {'ki': 0.4,  'storage': 0.020,  'max_drainage_ratio': 1},
        # Assume a simple extensive vegetated roof cover with 2" of growth media
        # at 15% porosity and 2" of granular discharge media at 25% porosity
        # Assume drainage area is equal only to the water that falls directly on the roof
        # (no extra contributing area).
        # Source:  PA stormwater manual 6.5.1
    'infiltration_basin':  {'ki': 0.0,  'storage': 0.610,  'max_drainage_ratio': 8},
        # Assume a large open area with no infiltration underneath and 2' of ponding depth (100% porosity)
        # Assume drainage area is largely pervious surface (lawns) allowing a maximum loading ratio of 8:1
        # Source:  New Jersey stormwater manual,  PA stormwater manual appendix C
    'porous_paving':        {'ki': 0.0,  'storage': 0.267,  'max_drainage_ratio': 2},
        # Assume porous bituminous asphalt used as the paving surface
        # 2.5" of porous paving service at 16% porosity, 1" of bedding layer/choker course at 50% porosity,
        # and 24" of infiltration bed/reservoir layer at 40% porosity
        # Assume some allowable additional drainage area (2:1) from roofs or adjacent pavement
        # Note that inflow from any pervious areas is not recommended due to potential clogging
        # Sources:  PA stormwater manual 6.4.1,
        # StormTech (http://www.stormtech.com/download_files/pdf/techsheet1.pdf),
        # http://www.construction.basf.us/features/view/pervious-pavements,
        # http: // stormwater.pca.state.mn.us / index.php / Design_criteria_for_permeable_pavement
    'rain_garden':          {'ki': 0.08, 'storage': 0.396,  'max_drainage_ratio': 5},
        # Assumes 6" of ponding depth at 100% porosity, 24" planting mix at 20% porosity
        # and 12" gravel underbed at 40% porosity.
        # Assume drainage area is largely impervious (parking lots) allowing a maximum loading ratio of 5:1
        # Source:  PA stormwater manual 6.4.5, PA stormwater manual appendix C
}

# Runoff tables for Pitt's Small Storm Hydrology (SSH) model

    # The raw runoff coefficients are those measured by the USGS in Wisconsin
    # (Bannerman 1983, 1992 and 1993; Horwatich, 2004; Steuer 1996 and 1997; USEPA 1993; Walker 1994; Waschbusch 1999)
    # This data is also provided as the Rv (runoff coefficient) file for all regions *but* the SouthEast in version 10.x of WinSlamm    #
    # http://wi.water.usgs.gov/slamm/index.html
    # http://wi.water.usgs.gov/slamm/slamm_parameter_descriptions.htm
    # http://winslamm.com/Select_documentation.html    #
    #
    # The Standard Land Uses, including the percents of land in area type and their level of connectedness,
    #  are collected from multiple published papers analyzing different sites using WinSLAMM
    # Pitt has compiled all of the site summaries here:
    # http://winslamm.com/docs/Standard%20Land%20Use%20and%20Parameter%20file%20descriptions%20final%20April%2018%202011.pdf
    # The above pdf also lists the original sources of the raw data.    #
    #
    # The final runoff volumens and runoff ratios for each standard land use were calculated as the sum of the multiples of the raw runoff coefficients
    # for each area type and the percent of land in that area type in each standard land use.
    #
    # For this work, this is the mapping used between the NLCD class and the SSH's Standard Land Use:
    #    NLCD class 21 (Developed, Open) = "Open Space"
    #    NLCD class 22 (Developed, Low) = "Residential"
    #    NLCD class 23 (Developed, Medium) = "Institutional"
    #    NLCD class 24 (Developed, High) = "Commercial"

    # The runoff coeffients for Cluster Housing were derived by taking the numbers for the residential SLU and
    # halving the amount of street, driveway, and parking and adding that amount to the amount of small
    # landscaping.  This simulates LID concentrating housing and decreasing paving, while maintaining the
    # same residential density (ie, the same amount of roof space).

SSH_RAINFALL_STEPS = [0.01, 0.08, 0.12, 0.2, 0.39, 0.59, 0.79, 0.98, 1.2, 1.6, 2, 2.4, 2.8, 3.2, 3.5, 3.9, 4.9]
SSH_RUNOFF_RATIOS = {
    'developed_open' :
         {'runoff_ratio':
              {'a': [0.0393, 0.0472, 0.0598, 0.0645, 0.1045, 0.1272, 0.1372, 0.1432, 0.1493, 0.1558, 0.1609, 0.1637, 0.1662, 0.1686, 0.1711, 0.1726, 0.1757],
               'b': [0.0393, 0.0472, 0.0598, 0.0645, 0.1177, 0.1462, 0.1636, 0.1697, 0.1809, 0.1874, 0.3127, 0.3148, 0.3165, 0.3182, 0.3199, 0.3206, 0.3229],
               'c': [0.0393, 0.0472, 0.0598, 0.0645, 0.1193, 0.1528, 0.1769, 0.1904, 0.2008, 0.2423, 0.3127, 0.3148, 0.3165, 0.3182, 0.3199, 0.3624, 0.4066],
               'd': [0.0393, 0.0472, 0.0598, 0.0645, 0.1193, 0.1528, 0.1769, 0.1904, 0.2008, 0.2423, 0.3127, 0.3148, 0.3165, 0.3182, 0.3199, 0.3624, 0.4066],
               }
         },
    'developed_low' :
         {'runoff_ratio':
              {'a' : [0.0785, 0.1115, 0.1437, 0.1601, 0.1841, 0.2053, 0.2138, 0.2187, 0.2249, 0.2303, 0.2359, 0.2382, 0.2412, 0.2439, 0.2465, 0.2485, 0.2523],
               'b' : [0.0785, 0.1115, 0.1437, 0.1601, 0.1960, 0.2224, 0.2377, 0.2426, 0.2534, 0.2589, 0.3731, 0.3748, 0.3770, 0.3791, 0.3809, 0.3822, 0.3853],
               'c' : [0.0785, 0.1115, 0.1437, 0.1601, 0.1974, 0.2284, 0.2496, 0.2614, 0.2714, 0.3085, 0.3731, 0.3748, 0.3770, 0.3791, 0.3809, 0.4200, 0.4609],
               'd' : [0.0785, 0.1115, 0.1437, 0.1601, 0.1974, 0.2284, 0.2496, 0.2614, 0.2714, 0.3085, 0.3731, 0.3748, 0.3770, 0.3791, 0.3809, 0.4200, 0.4609],
               }
         },
    'developed_med' :
         {'runoff_ratio':
              {'a' : [0.1322, 0.1929, 0.2631, 0.3107, 0.3698, 0.4032, 0.4235, 0.4368, 0.4521, 0.4688, 0.4816, 0.4886, 0.4953, 0.5006, 0.5047, 0.5074, 0.5138],
               'b' : [0.1322, 0.1929, 0.2631, 0.3150, 0.3838, 0.4226, 0.4474, 0.4616, 0.4797, 0.4980, 0.5715, 0.5803, 0.5887, 0.5944, 0.6002, 0.6045, 0.6146],
               'c' : [0.1322, 0.1929, 0.2631, 0.3150, 0.3846, 0.4258, 0.4539, 0.4717, 0.4895, 0.5249, 0.5715, 0.5803, 0.5887, 0.5944, 0.6002, 0.6248, 0.6553],
               'd' : [0.1322, 0.1929, 0.2631, 0.3150, 0.3846, 0.4258, 0.4539, 0.4717, 0.4895, 0.5249, 0.5715, 0.5803, 0.5887, 0.5944, 0.6002, 0.6248, 0.6553],
               }
         },
    'developed_high' :
         {'runoff_ratio':
              {'a': [0.1966, 0.2815, 0.4034, 0.4796, 0.5549, 0.6037, 0.6311, 0.6471, 0.6675, 0.6891, 0.7063, 0.7154, 0.7257, 0.7335, 0.7389, 0.7435, 0.7533],
               'b': [0.1966, 0.2815, 0.4034, 0.4895, 0.5803, 0.6343, 0.6647, 0.6818, 0.7045, 0.7274, 0.7724, 0.7820, 0.7925, 0.8005, 0.8059, 0.8104, 0.8203],
               'c': [0.1966, 0.2815, 0.4034, 0.4895, 0.5807, 0.6358, 0.6677, 0.6865, 0.7090, 0.7398, 0.7724, 0.7820, 0.7925, 0.8005, 0.8059, 0.8197, 0.8390],
               'd': [0.1966, 0.2815, 0.4034, 0.4895, 0.5807, 0.6358, 0.6677, 0.6865, 0.7090, 0.7398, 0.7724, 0.7820, 0.7925, 0.8005, 0.8059, 0.8197, 0.8390],
               }
         },
    'cluster_housing' :
         {'runoff_ratio':
              {'a': [0.0466, 0.0733, 0.0956, 0.1084, 0.1262, 0.1387, 0.1452, 0.1492, 0.1538, 0.1580, 0.1623, 0.1641, 0.1664, 0.1684, 0.1701, 0.1717, 0.1743],
               'b': [0.0466, 0.0733, 0.0956, 0.1084, 0.1395, 0.1578, 0.1718, 0.1758, 0.1856, 0.1897, 0.3146, 0.3157, 0.3171, 0.3183, 0.3193, 0.3201, 0.3218],
               'c': [0.0466, 0.0733, 0.0956, 0.1084, 0.1411, 0.1645, 0.1851, 0.1966, 0.2056, 0.2449, 0.3146, 0.3157, 0.3171, 0.3183, 0.3193, 0.3619, 0.4056],
               'd': [0.0466, 0.0733, 0.0956, 0.1084, 0.1411, 0.1645, 0.1851, 0.1966, 0.2056, 0.2449, 0.3146, 0.3157, 0.3171, 0.3183, 0.3193, 0.3619, 0.4056],
               }
         },
}


# The set of best management practices that we know about.  The
# cluster_housing and no_till types are excluded because they do not
# actively retain water.
BMPS = set(['green_roof', 'porous_paving',
            'rain_garden', 'infiltration_basin'])

# The set of "built" land uses
# These are the land uses to which the Pitt model will be applied at less than 2" of rain.
BUILT_TYPES = set(['developed_open', 'developed_low', 'developed_med',
                   'developed_high', 'cluster_housing'])

NON_NATURAL = set(['pasture', 'cultivated_crops', 'green_roof']) | set(['no_till']) | BMPS | BUILT_TYPES

# The set of pollutants that we are concerned with.
POLLUTANTS = set(['tn', 'tp', 'bod', 'tss'])

# Event mean concentrations (mg/l) by pollutant and NLCD type
# tn: Total Nitrogen, tp: Total Phosphorus,
# bod: Biochemical Oxygen Demand, tss: Total Suspended Solids
# Data from:
# (1)	USEPA, 2011. User’s Guide: Spreadsheet Tool for Estimation of Pollutant Load (STEPL), Version 4.1, 57 pp.
# (2)	Pennsylvania Department of Environmental Protection, 2006.
#       Pennsylvania Stormwater Best Management Practices Manual. 685 pp.
# (3)	USEPA, 2005. The National Stormwater Quality Database, Version 1.2: A Compilation and Analysis of NPDES
#       Stormwater Monitoring Information. USEPA, Office of Water, Washington, DC, 447 pp.
# (4)	New Hampshire Dept. of Environmental Services, 2010. Guidance for Estimating Pre- and Post-Development
#       Stormwater Loads. (EMCs available at
#       http://www.des.nh.gov/organization/divisions/water/stormwater/documents/wd-08-20a_apxd.pdf)
# (5)	Washington State Dept. of Ecology, 2007. Efficiency of Urban Stormwater Best Management Practices:
#       A Literature Review.  Publication No. 07-03-009, 12 pp.
# (6)	Keiser & Associates, 2003. Empirical Sediment and Phosphorus Nonpoint Source Model for the St. Joseph River
#       Watershed. 48 pp.
POLLUTION_LOADS = {
    11: {'tn': 0,    'tp': 0,     'bod': 0,     'tss': 0},      # Open water
    12: {'tn': 0,    'tp': 0,     'bod': 0,     'tss': 0},      # Perennial Ice/Snow
    21: {'tn': 2.26, 'tp': 0.32,  'bod': 5.7,   'tss': 99.8},   # Developed, Open Space
    22: {'tn': 2.58, 'tp': 0.38,  'bod': 6.0,   'tss': 126},    # Developed, Low Intensity
    23: {'tn': 3.62, 'tp': 0.38,  'bod': 9.0,   'tss': 134.7},  # Developed, Medium Intensity
    24: {'tn': 3.54, 'tp': 0.35,  'bod': 9.9,   'tss': 163.7},  # Developed High Intensity
    31: {'tn': 0.10, 'tp': 0.01,  'bod': 0.0,   'tss': 1},      # Barren Land (Rock/Sand/Clay)
    32: {'tn': 0.10, 'tp': 0.01,  'bod': 0.0,   'tss': 1},      # Quarries/Strip Mines/Gravel Pits
    41: {'tn': 1.05, 'tp': 0.13,  'bod': 0.5,   'tss': 45},     # Deciduous Forest
    42: {'tn': 1.05, 'tp': 0.13,  'bod': 0.5,   'tss': 45},     # Evergreen Forest
    43: {'tn': 1.05, 'tp': 0.13,  'bod': 0.5,   'tss': 45},     # Mixed Forest
    51: {'tn': 0,    'tp': 0,     'bod': 0,     'tss': 0},      # Dwarf Scrub (Alaska Only, N/A)
    52: {'tn': 0.19, 'tp': 0.15,  'bod': 0.5,   'tss': 39},     # Shrub/Scrub
    71: {'tn': 2.30, 'tp': 0.22,  'bod': 0.5,   'tss': 48.8},   # Grassland/Herbaceous
    72: {'tn': 0,    'tp': 0,     'bod': 0,     'tss': 0},      # Sedge/Herbaceous (Alaska Only, N/A)
    73: {'tn': 0,    'tp': 0,     'bod': 0,     'tss': 0},      # Lichens (Alaska Only, N/A)
    74: {'tn': 0,    'tp': 0,     'bod': 0,     'tss': 0},      # Moss (Alaska Only, N/A)
    81: {'tn': 5.71, 'tp': 0.55,  'bod': 13,    'tss': 145},    # Pasture/Hay
    82: {'tn': 7.70, 'tp': 1.07,  'bod': 12.45, 'tss': 216},    # Cultivated Crops
    90: {'tn': 0.19, 'tp': 0.006, 'bod': 0.5,   'tss': 0},      # Woody Wetlands
    91: {'tn': 0.19, 'tp': 0.006, 'bod': 0.5,   'tss': 0},
    92: {'tn': 0.19, 'tp': 0.006, 'bod': 0.5,   'tss': 0},
    93: {'tn': 0.19, 'tp': 0.006, 'bod': 0.5,   'tss': 0},
    94: {'tn': 0.19, 'tp': 0.006, 'bod': 0.5,   'tss': 0},
    95: {'tn': 0.19, 'tp': 0.006, 'bod': 0.5,   'tss': 0},      # Emergent Herbaceous Wetlands
    96: {'tn': 0.19, 'tp': 0.006, 'bod': 0.5,   'tss': 0},
    97: {'tn': 0.19, 'tp': 0.006, 'bod': 0.5,   'tss': 0},
    98: {'tn': 0.19, 'tp': 0.006, 'bod': 0.5,   'tss': 0},
    99: {'tn': 0.19, 'tp': 0.006, 'bod': 0.5,   'tss': 0}
}
